File: web_sdk/src/__tests__/key-generator.test.ts
// src/__tests__/key-generator.test.ts

import { describe, it, expect } from 'vitest';
import { KeyGenerator } from '../crypto/key-generator.js';
import { exportPublicKeyRaw, verifySPKSignature } from '../crypto/crypto-utils.js';

describe('KeyGenerator', () => {
  it('generateKeyBundle produces valid bundle', async () => {
    const bundle = await KeyGenerator.generateKeyBundle(5);
    expect(bundle.otpks).toHaveLength(5);
    expect(bundle.nextOtpkId).toBe(6);
    expect(bundle.signedPrekey.id).toBe(1);
  });

  it('SPK signature self-verifies', async () => {
    const bundle = await KeyGenerator.generateKeyBundle(1);
    const spkPub = await exportPublicKeyRaw(bundle.signedPrekey.keyPair.publicKey);
    const valid = await verifySPKSignature(
      spkPub,
      bundle.signedPrekey.signature,
      bundle.identityKeyEd.publicKey,
    );
    expect(valid).toBe(true);
  });

  it('OTPKs have sequential unique IDs', async () => {
    const otpks = await KeyGenerator.generateOtpks(10, 5);
    expect(otpks).toHaveLength(10);
    const ids = otpks.map((o) => o.id);
    expect(ids).toEqual([5, 6, 7, 8, 9, 10, 11, 12, 13, 14]);
  });

  it('all OTPKs have distinct public keys', async () => {
    const otpks = await KeyGenerator.generateOtpks(5, 1);
    const pubKeys = await Promise.all(
      otpks.map((o) => exportPublicKeyRaw(o.keyPair.publicKey).then((b) => Buffer.from(b).toString('hex'))),
    );
    const unique = new Set(pubKeys);
    expect(unique.size).toBe(5);
  });

  it('generateSenderKey produces 32-byte chain key', async () => {
    const sk = await KeyGenerator.generateSenderKey();
    expect(sk.chainKey.byteLength).toBe(32);
    expect(sk.signingKey.publicKey).toBeDefined();
    expect(sk.signingKey.privateKey).toBeDefined();
  });
});

File: flutter_sdk/test/crypto/key_generator_test.dart
import 'package:test/test.dart';
import 'package:rust_e2e_chat_sdk/src/crypto/key_generator.dart';

void main() {
  group('KeyGenerator', () {
    test('generateKeyBundle produces correct key counts', () async {
      final bundle = await KeyGenerator.generateKeyBundle(otpkCount: 10);
      expect(bundle.identityKeyPair.publicKey.length, equals(32));
      expect(bundle.identityKeyPair.privateKey.length, equals(32));
      expect(bundle.signedPreKeyPair.publicKey.length, equals(32));
      expect(bundle.signedPreKeySignature, isNotEmpty);
      expect(bundle.oneTimePreKeys.length, equals(10));
    });

    test('all OTPKs have unique IDs', () async {
      final bundle = await KeyGenerator.generateKeyBundle(otpkCount: 20);
      final pubKeys =
          bundle.oneTimePreKeys.map((k) => k.keyPair.publicKey.join(',')).toSet();
      expect(pubKeys.length, equals(20));
    });

    test('SPK signature verifies correctly', () async {
      final bundle = await KeyGenerator.generateKeyBundle(otpkCount: 0);
      final valid = await KeyGenerator.verifySignature(
        data: bundle.signedPreKeyPair.publicKey,
        signature: bundle.signedPreKeySignature,
        publicKey: bundle.identitySigningKeyPair.publicKey,
      );
      expect(valid, isTrue);
    });

    test('tampered SPK signature fails verification', () async {
      final bundle = await KeyGenerator.generateKeyBundle(otpkCount: 0);
      final tampered = List<int>.from(bundle.signedPreKeySignature);
      tampered[0] ^= 0xFF; // flip bits
      final valid = await KeyGenerator.verifySignature(
        data: bundle.signedPreKeyPair.publicKey,
        signature: tampered as dynamic,
        publicKey: bundle.identitySigningKeyPair.publicKey,
      );
      expect(valid, isFalse);
    });

    test('generateSenderKey produces non-empty chain key', () async {
      final pair = await KeyGenerator.generateSenderKey();
      expect(pair.chainKey.length, equals(32));
      expect(pair.signingKeyPair.publicKey.length, equals(32));
      expect(pair.signingKeyPair.privateKey.length, equals(32));
    });

    test('two generateKeyBundle calls produce different keys', () async {
      final b1 = await KeyGenerator.generateKeyBundle(otpkCount: 0);
      final b2 = await KeyGenerator.generateKeyBundle(otpkCount: 0);
      expect(
        b1.identityKeyPair.publicKey,
        isNot(equals(b2.identityKeyPair.publicKey)),
      );
    });
  });
}
